File: oracle/ghidra/DllSurvey.py
# DllSurvey.py -- survey eloquence-family romanizer binaries.

# Prints to scriptlog. Writes <bin>.strings.txt (rom-block strings(

# and <bin>.survey.decomp.txt (C of top-15 prober functions(.

from ghidra.program.model.address import AddressSet

from ghidra.app.decompiler import DecompInterface

from ghidra.util.task import ConsoleTaskMonitor


def ro_name(b) : return str(b.getName()).lower()


def is_ro_data(b) :

    if not b.isRead() : return False

    if b.isWrite() : return False

    if not b.isInitialized() : return False

    bn = ro_name(b)

    return "rdata" in bn or "const" in bn or "rodata" in bn or "cstring" in bn or "data.rel.ro" in bn


def pick_ro_data(blks) :

    best = None

    best_size = -1

    for b in blks :

        if is_ro_data(b)and b.getSize() > best_size :

            best = b

            best_size = b.getSize()

    return best


def pick_largest(blks) :

    best = None

    best_size = -1

    for b in blks :

        if b.getSize() > best_size :

            best = b

            best_size = b.getSize()

    return best
p = getCurrentProgram()

print("PROGRAM " + p.getName() + " arch=" + p.getLanguage().getLanguageID())

blks = p.getMemory().getBlocks()

for b in blks :

    flags = ""

    if b.isRead() : flags += "R"

    if b.isWrite() : flags += "W"

    if b.isExecute() : flags += "X"

    print("SECTION %-16s size=0x%x flags=%s init=%s" % (ro_name(b) , b.getSize() , flags , b.isInitialized()) )

largestRO = pick_ro_data(blks)



if largestRO is None : largestRO = pick_largest(blks)



mem = p.getMemory()

start = largestRO.getStart()

size = largestRO.getSize()

cur = []

samples = []

str_count = 0

for i in xrange(int(size)) :

    v = mem.getByte(start.add(i))

    if 0x20 <= v < 0x7f : cur.append(chr(v))



    else :

        if len(cur) >= 4 :

            s = "".join(cur)

            str_count = str_count + 1

            if len(samples) < 8 : samples.append(s[:160])

        cur = []

strings_path = "/tmp/dllsurv/%s.strings.txt" % p.getName()

w = open(strings_path, "w")

w.write("# strings in %s (count=%d) samples:\n" % (largestRO.getName(), str_count))

for s in samples : w.write(s + "\n")

w.close()

print("STRINGS count=%d samples=%d -> %s" % (str_count , len(samples) , strings_path))

listing = p.getListing()

rm = p.getReferenceManager()

roset = AddressSet(largestRO.getStart(), largestRO.getEnd())

refcount = {}

it = rm.getReferenceIterator(roset)



while it.hasNext() :

    from_add = it.next().getFromAddress()

    f = listing.getFunctionContaining(from_add)





    if f is not None : refcount[f] = refcount.get(f, 0) + 1

print("=== PROBERS top-15 referencing %s:" % largestRO.getName())

items = sorted(refcount.items(), key=lambda kv : -kv[1])

for fobj , cnt in items[:15] : print("  %s refs=%d" % (fobj.getName() , cnt))

dec = DecompInterface()

dec.toggleCCode(True)

dec.toggleSyntaxTree(True)



if not dec.openProgram(p) :
    print("DecompInterface failed: " + dec.getLastMessage())

elif len(items) ==0 : print("DECOMP no probers found")

else :

    decomp_path = "/tmp/dllsurv/%s.survey.decomp.txt" % p.getName()

    out = open(decomp_path, "w")

    n =  0

    for fobj, cnt in items[:15] :

        res = dec.decompileFunction(fobj, 60, ConsoleTaskMonitor())

        if res is not None and res.decompileCompleted() :

            out.write("// ===== %s refs=%d\n%s\n" % (fobj.getName(), cnt, res.getDecompiledFunction().getC()))

            n = n + 1

        else : out.write("// ===== %s refs=%d  [decomp failed]\n" % (fobj.getName() , cnt))

    out.close()

    print("DECOMP wrote %d functions -> %s" % (n , decomp_path))

dec.dispose()

File: oracle/ghidra/DllSurvey.java
// DllSurvey.java — Ghidra headless survey for Eloquence-family rom DLLs.
// Usage: analyzeHeadless <proj> DllSurvey -import <file> -scriptPath <dir> -postScript DllSurvey.java -scriptlog <log>
// Outputs to stdout (captured via -scriptlog):
//   1) Section table (name, size, flags)
//   2) String inventory in rom-bearing sections (.rdata/.data/.m2e_*): count + first 40 samples
//   3) Top functions by number of references INTO the largest read-only section (rom probers)
//   4) Decompilation of those functions appended to <input>.survey.decomp.txt
import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.*;
import ghidra.program.model.mem.*;
import ghidra.program.model.address.*;
import ghidra.program.model.symbol.*;
import ghidra.app.decompiler.*;
import java.io.*;
import java.util.*;

public class DllSurvey extends GhidraScript {
    @Override
    public void run() throws Exception {
        Program p = currentProgram;
        println("=== PROGRAM " + p.getName() + " arch=" + p.getLanguage().getLanguageID());

        // 1) sections
        MemoryBlock[] blks = p.getMemory().getBlocks();
        MemoryBlock largestRO = null;
        for (MemoryBlock b : blks) {
            String flags = (b.isRead() ? "R" : "") + (b.isWrite() ? "W" : "") + (b.isExecute() ? "X" : "");
            println(String.format("SECTION %-16s size=0x%x flags=%s init=%s", b.getName(), b.getSize(), flags, b.isInitialized()));
            String bn = b.getName().toLowerCase();
            boolean dataHunk = bn.contains("rdata") || bn.contains("const") || bn.contains("rodata") || bn.contains("cstring") || bn.contains("data.rel.ro");
            if (b.isRead() && !b.isWrite() && b.isInitialized() && dataHunk && (largestRO == null || b.getSize() > largestRO.getSize()))
                largestRO = b;
        }
        if (largestRO == null) {
            for (MemoryBlock b : blks) {
                String bn = b.getName().toLowerCase();
                boolean dataHunk = bn.contains("rdata") || bn.contains("const") || bn.contains("rodata") || bn.contains("cstring") || bn.contains("data.rel.ro");
                if (b.isRead() && !b.isWrite() && b.isInitialized() && dataHunk && (largestRO == null || b.getSize() > largestRO.getSize()))
                    largestRO = b;
            }
        }
        if (largestRO == null) {
            for (MemoryBlock b : blks) {
                if (b.isRead() && !b.isWrite() && b.isInitialized() && (largestRO == null || b.getSize() > largestRO.getSize()))
                    largestRO = b;
            }
        }

        // 2) strings in initialized non-exec sections
        if (largestRO != null) {
            Memory mem = p.getMemory();
            Address start = largestRO.getStart();
            long size = largestRO.getSize();
            StringBuilder cur = new StringBuilder();
            List<String> samples = new ArrayList<>();
            long strCount = 0;
            for (long i = 0; i < size; i++) {
                byte v = mem.getByte(start.add(i));
                if (v >= 0x20 && v < 0x7f) { cur.append((char) v); }
                else {
                    if (cur.length() >= 4) {
                        strCount++;
                        if (samples.size() < 40) samples.add(start.add(i - cur.length()) + " len=" + cur.length() + "  " + cur.substring(0, Math.min(cur.length(), 60)));
                    }
                    cur.setLength(0);
                }
            }
            println(String.format("ROSTRINGS section=%s count=%d (len>=4)", largestRO.getName(), strCount));
            for (String s : samples) println("  " + s);
        }

        // 3) functions referencing the largest RO block (rom probers)
                if (largestRO != null) {
                    Listing listing = p.getListing();
                    Map<Function, Integer> refcount = new LinkedHashMap<>();
                    ReferenceManager rm = p.getReferenceManager();
                    AddressSet roset = new AddressSet(largestRO.getStart(), largestRO.getEnd());
                    AddressIterator it = rm.getReferenceSourceIterator(roset, true);
                                        while (it.hasNext()) {
                                            Address from =it.next();
                        Function f = listing.getFunctionContaining(from);
                        if (f != null) {
                            refcount.put(f, refcount.getOrDefault(f, 0) + 1);
                        }
                    }
                    List<Map.Entry<Function, Integer>> sorted = new ArrayList<>(refcount.entrySet());
                    sorted.sort((a, b) -> b.getValue() - a.getValue());
                    println("=== PROBERS top-15 referencing " + largestRO.getName() + ":");
                    int shown = 0;
                    for (Map.Entry<Function, Integer> e : sorted) {
                        if (shown++ >= 15) break;
                        println("  " + e.getKey().getName() + " refs=" + e.getValue());
                    }
                    // 4) decompile them
                    DecompInterface dec = new DecompInterface();
                    dec.toggleCCode(true); dec.toggleSyntaxTree(true);
                    if (!dec.openProgram(p)) { println("DecompInterface failed: " + dec.getLastMessage()); return; }
                    File out = new File("/tmp/dllsurv/" + p.getName() + ".survey.decomp.txt");
                    PrintWriter w = new PrintWriter(new FileWriter(out));
                    shown =0;
                    for (Map.Entry<Function, Integer> e : sorted) {
                        if (shown++ >= 15) break;
                        Function f = e.getKey();
                if (f == null) continue;
                DecompileResults res = dec.decompileFunction(f, 60, monitor);
                w.println("// ===== " + e.getKey() + " refs=" + e.getValue());
                w.println(res.getDecompiledFunction() == null ? "// FAILED " + dec.getLastMessage() : res.getDecompiledFunction().getC());
                w.flush();
            }
            w.close();
            println("DECOMP wrote " + out.getAbsolutePath());
            dec.dispose();
        }
    }
}